### PR TITLE
Improve test coverage in the catboost integration tests

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Follow up of #17931. This PR tests catboost with categorical features and increases the size of the the test data we fit our catboost models on.
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
@jameslamb you had some thoughts on the last PR, any interest in taking another stab at reviewing this one ;)

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py [hunk 1]
```diff
@@ -84,24 +95,26 @@ def test_catboost_with_pool_and_dataframe(regression_data):
...
 
 
 def test_catboost_with_categorical_features():
     data = {
-        "numerical_feature": rng.standard_normal(100),
-        "categorical_feature": rng.choice(["A", "B", "C"], size=100),
-        "target": rng.integers(0, 2, size=100),
+        "fea_1": rng.standard_normal(1000),
+        "fea_2": rng.choice([0, 1, 2, 3], size=1000),
+        "fea_3": rng.choice(["A", "B", "C"], size=1000),
```
Comment (jameslamb) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py:108]:
```suggestion
        "cat_feat_int_dtype": rng.choice([0, 1, 2, 3], size=1000),
        "cat_feat_categorical_dtype": rng.choice(["A", "B", "C"], size=1000),
```

Would you consider a change like this, or a code comment capturing this information if you'd prefer? Something really important is happening here...

* `fea_2` is testing "using an `int` column as a categorical feature works"
* `fea_3` is testing "using a `categorical` column as a categorical feature works"

*(based on https://github.com/rapidsai/cudf/pull/17931#discussion_r1947207063)*

But I don't think that important difference would be obvious to someone looking at this test for the first time who isn't familiar with these boosting libraries.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py [hunk 2]
```diff
@@ -84,24 +95,26 @@ def test_catboost_with_pool_and_dataframe(regression_data):
...
     return predictions
 
 
 def test_catboost_with_categorical_features():
     data = {
-        "numerical_feature": rng.standard_normal(100),
-        "categorical_feature": rng.choice(["A", "B", "C"], size=100),
-        "target": rng.integers(0, 2, size=100),
+        "fea_1": rng.standard_normal(1000),
+        "fea_2": rng.choice([0, 1, 2, 3], size=1000),
```
Comment (jameslamb) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py:107]: Choosing these values randomly like this is a risk. The categorical features may be noise to `catboost`, meaning they wouldn't be chosen for any splits. And if they aren't chosen for any splits, then this test wouldn't be able to catch bugs in encoding/decoding that lead to the wrong representation (because `predict()` would effectively ignore the values in these columns).

It's for concerns like this that the `xgboost` example code I shared tries to create categorical features that have some relationship to the target: https://github.com/dmlc/xgboost/blob/105aa4247abb3ce787be2cef2f9beb4c24b30049/demo/guide-python/categorical.py#L40-L45

For these tests, I think the easiest way to do that would be something like this:

* use `sklearn.datasets.make_classification()` to generate the initial dataset (features there are not totally random... they have some relationship to the target)
* to get an `int` categorical feature, replace one of those continuous columns with a discretized version (say, split it into deciles with `np.quantile()` and `np.digitize()` or something)... e.g. values in the lowest 10% coded a `0`, next 10-20% as `1`, etc.
* to get a `category` feature, repeat that procedure (but with a different feature's values, and using a different cardinality), but instead of ints use strings like `"low"`, `"medium"`, `"high"` (or whatever you want) and convert to `category` dtype

And then in the test, assert that all the features were used at least once. I don't know the exact API for this in `catboost`... in `xgboost` / `lightgbm` it is a property on the estimator called `.feature_importances_`, like in this example: https://stackoverflow.com/a/79435055/3986677

I know this seems like a lot, but I think it's important. When given a `pandas`-like input to `.predict()`, `catboost` will try to do the same types of transformation as it did on the training data. There's potential for bugs in this integration there, that'd only be caught at `predict()` time by ensuring the model actually uses the categorical features.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py [hunk 3]
```diff
@@ -112,17 +125,17 @@ def test_catboost_with_categorical_features():
...
-        (rng.standard_normal((100, 5)), rng.standard_normal(100)),
+        (rng.standard_normal((1000, 5)), rng.standard_normal(1000)),
     ],
 )
 def test_catboost_train_test_split(X, y):
     from sklearn.model_selection import train_test_split
 
     X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
-    model = CatBoostRegressor(iterations=10, verbose=0)
+    model = CatBoostRegressor(iterations=50, verbose=0)
```
Comment (jameslamb) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py:138]:
```suggestion
    model = CatBoostRegressor(iterations=50, random_state=708, verbose=0)
```

For all of these tests, you should pass in a non-0 seed to control the randomness internal to `catboost`. I'm not certain if `catboost` training is deterministic given the parameters you've supplied, but this is still a good idea just to be safe and reduce the risk of flakiness in these tests.

**Review by jameslamb — COMMENTED** [review]
Thanks for the `@`, and for doing this follow-up PR!

I left a few more suggestions for your consideration. Leaving a non-blocking review... they're all things that shouldn't block a merge and that you can choose whether or not to do based on how much time you want to invest in this.

**Review by trivialfis — COMMENTED** [review]
Not familiar with catboost, overall, it would be useful to have tests that check the consistency of the predict function as James suggested. For instance, what happens if the test dataset has lesser categories. What happens if there's null values. What happens if the test dataset has wrong categories, etc. Just to make sure catboost is producing the same results as pandas inputs.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py (hunk 2 above)
Comment (Matt711) [reply python/cudf/cudf_pandas_tests/third_party_integration_tests/tests/test_catboost.py]: > to get a category feature, repeat that procedure (but with a different feature's values, and using a different cardinality), but instead of ints use strings like "low", "medium", "high" (or whatever you want) and convert to category dtype

I was unable to add this feature. When doing so I ran into this closed catboost issue https://github.com/catboost/catboost/issues/934 which in IMO should still be open since several other have run into it.

**Review by vyasr — APPROVED** [review]
Looks like a solid improvement. I'll trust the other reviewers expertise in judging if this is a sufficient test for what we need to verify works with catboost.
